Add first version of figure layout

sophy-korp opened the pull request:
Вначале идёт работа с константами. Определяем размеры холста, минимальное расстояние между фигурами, размеры объектов по умолчанию. Определяются структуры для хранения информации об объектах и связях между ними.
Функция layout_objects() - основная функция расстановки. Она работает примерно так:
1. Разбирает программу на объекты и связи, а именно: создаёт списки вышеупомянутых структур objects и relations; вызывает функцию analyze_program, которая заполняет эти списки.
2. Разделяет объекты на группы связанных объектов = компоненты (связности), а именно: создаёт список списков объектов = список компонент; вызывает find_connected_components для поиска связанных объектов, которая использует поиск в глубину.
3. Распределяет компоненты по холсту. (Находим связанные компоненты и распределяем по холсту их, а не отдельные фигуры, потому что так связанные объекты будут находиться рядом и меньше вероятность нежелательных пересечений связей). Компоненты размещаются слева направо, сверху вниз. Есть проверка, помещается ли компонента в текущей строке, если нет, то она размещается в следующей строке. Есть функция расположения объектов внутри самой компоненты - layout_component. Она использует обход в ширину и располагает соседей справа от текущего объекта с учётом размеров. Думаю, возможно, эту функцию можно адаптировать для применения отдельно к графам.
4. Дальше идёт работа со связями, мы говорили, что, может быть, эту часть стоит убрать. Вызывается generate_relation_points, которая генерирует точки, через которые нужно провести связь. В зависимости от того, какое расстояние между фигурами больше, вертикальное или горизонтальное, она определяет, соединить их за боковые стороны или за дно и верх. Затем создаёт путь из 3 участков.

sophy-korp committed:
Add files via upload

11AgReS1SoR11 commented on Docs/Расположение фигур.txt:
```diff
@@ -0,0 +1,330 @@
+        component_width, component_height = calculate_component_size(component, relations)
+        
+        # Проверяем, помещается ли компонента в текущей строке
+        IF current_x + component_width > CANVAS_WIDTH:
```
Мне кажется или здесь нужен while?

sophy-korp replied:
Мне не кажется, что нужен, это внутри for по компонентам. Может, я не понимаю идею замечания и поэтому не вижу, что есть ошибка

11AgReS1SoR11 commented on Docs/Расположение фигур.txt:
```diff
@@ -0,0 +1,330 @@
+    DECLARE queue AS QUEUE OF ObjectInfo = [root]
+    DECLARE current_level AS INTEGER = 0
+    
+    WHILE queue IS NOT EMPTY:
```
А где извлечение из очереди? Кажется, что это бесконечный цикл?

sophy-korp replied:
Там где DEQUEUE(queue), которое должно быть pop

11AgReS1SoR11 commented on Docs/Расположение фигур.txt:
```diff
@@ -0,0 +1,330 @@
+        DECLARE max_height_in_level AS INTEGER = 0
+        
+        FOR i FROM 0 TO level_size - 1:
+            DECLARE obj AS ObjectInfo = DEQUEUE(queue)
```
А что такое DEQUEUE?

sophy-korp replied:
Тут должно было быть pop. Извлечение первого элемента

11AgReS1SoR11 commented on Docs/Расположение фигур.txt:
```diff
@@ -0,0 +1,330 @@
+            start_point = calculate_connection_point(obj1, start_side)
+            end_point = calculate_connection_point(obj2, end_side)
+            
+            IF abs(start_point.x - end_point.x) < 50:
```
50 - это что за константа?

sophy-korp replied:
Это произвольная константа, которую мы можем выбрать. Он нужна для того, чтобы оценить, насколько близко друг к другу объекты по x. Если они достаточно близко, то изламывать линию соединения не нужно

11AgReS1SoR11 replied:
Для этого надо бы отдельную константу и как-то её назвать, а то бесхозное число какое-то

11AgReS1SoR11 reviewed (commented):
Тяжёлый алгоритм, я мельком пробежался и ничего не понял. Позже рассмотрю более подробно. Тебе стоит в описании MR хотя бы идею рассказать, а то тяжёло вникнуть по коду

sophy-korp committed:
Update Расположение фигур.txt

Объявлена отдельная константа для проверки того, насколько близко расположены друг к другу фигуры. DEQUEUE исправлено на POP

sophy-korp committed:
Update Расположение фигур.txt

Расположение фигур + расположение графов + расположение dot_cloud

11AgReS1SoR11 commented on Docs/Расположение фигур.txt:
```diff
@@ -0,0 +1,381 @@
+FUNCTION layout_component(component, relations, start_x, start_y):
+    # Если в компоненте только один объект
+    IF LENGTH(component) == 1:
+        component[0].x = start_x
```
[major] Вот здесь логическая ошибка. Проперти X и Y у компонент - это координаты ЦЕНТРА. А ты передаёшь левую границу компоненты. Это надо учитывать, когда задаёшь координаты фигурам и сдигать центра на половину ширины фигуры

11AgReS1SoR11 commented on Docs/Расположение фигур.txt:
```diff
@@ -0,0 +1,381 @@
+            FOR rel IN relations:
+                IF rel.id1 == obj.id AND rel.id2 NOT IN visited:
+                    DECLARE neighbor AS ObjectInfo = FIND_OBJECT_BY_ID(rel.id2, objects)
+                    IF neighbor:
```
[minor] А это что за проверка? Такого объекта может не быть? Кажется, что мы такое проверили на предыдущем шаге семантического анализатора (Ниже аналогично)

11AgReS1SoR11 commented on Docs/Расположение фигур.txt:
```diff
@@ -0,0 +1,381 @@
+                FOR neighbor IN neighbors:
+                    total_neighbors_width += neighbor.width + DEFAULT_SPACING
+                
+                current_y = start_y + (MAX_SHAPE_HEIGHT() - total_neighbors_width)/2
```
[вопрос] А что это за формула? Из высоты вычитаем ширину? Это почему?

11AgReS1SoR11 commented on Docs/Расположение фигур.txt:
```diff
@@ -0,0 +1,381 @@
+                
+                FOR neighbor IN neighbors:
+                    neighbor.x = current_x
+                    neighbor.y = current_y
```
[вопрос] Т е твой алгоритм располагает компоненты связаности по горизонтали, а внутри этих компонент располагает объекты по вертикале?

11AgReS1SoR11 commented on Docs/Расположение фигур.txt:
```diff
@@ -0,0 +1,381 @@
+                    max_height_in_level = MAX(max_height_in_level, neighbor.height)
+        
+        current_level += 1
+        start_y += max_height_in_level + DEFAULT_SPACING
```
Ну и та же ошибка. У тебя эта переменная даже называется start_y - откуда начинаем, но при обращении к neighbor.x = нужно учитывать размеры этой фигуры

11AgReS1SoR11 commented on Docs/Расположение фигур.txt:
```diff
@@ -0,0 +1,381 @@
+            # Определяем стороны соединения в зависимости от относительного положения
+            DECLARE start_side, end_side AS STRING
+
+            IF abs(obj1.y - obj2.y) > abs(obj1.x - obj2.x):
```
[вопрос] Это что за проверка? Опять я не понимаю, почему мы сравниваем расстояние центров по вертикали с расстоянием центров по горизонтали?

Upd: насколько я помню, эту функцию мы решили не делать, потому что полагаемся на фронтенд в расположении линий? Если да, то убирай эту функцию, чтобы не запутаться

11AgReS1SoR11 commented on Docs/Расположение фигур.txt:
```diff
@@ -0,0 +1,381 @@
+SMALL_DISTANCE = 50 # константа для проверки близкого расположения фигур
+
+DECLARE SHAPE_SIZES AS MAP OF SHAPE TO (width, height)
+SHAPE_SIZES["circle"] = (100, 100)    # Диаметр круга по умолчанию
```
И тут не понятно, что значит (100, 100) с комментарием "диаметр". Диаметр 2 числами задаётся? Я так понимаю, что это квадрат, в который вписан этот круг, потому что так удобнее с ним работать в твоём алгоритме будет, я думаю.

Это относится и к ромбу. У ромба стороны задаются с углом между ними. Соответственно имеет смысл засунуть его во внутрь проямоугольника и работать дальше с этим, как с прямоугольником

sophy-korp committed:
Update Расположение фигур.txt

sophy-korp committed:
Update Расположение фигур.txt
